File: leetcode/278-First-Bad-Version/FirstBadVersion_001.java
/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        int l = 1, r = n;
        while (l <= r) {
            int m = l + (r - l) / 2; // key point: avoid overflow which causes strange behavior of m
            if (isBadVersion(m)) {
                r = m - 1;
            } else {
                l = m + 1;
            }
        }
        
        return l;
    }
}

File: leetcode/278-First-Bad-Version/FirstBadVersion_001.py
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return a bool
# def isBadVersion(version):
#     target = 3
#     return version >= target

class Solution(object):
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        l, r = 1, n # l, r = 0, n - 1 also works
        while l <= r:
            m = (l + r) / 2
            if isBadVersion(m):
                r = m - 1
            else:
                l = m + 1
        return l
